src/mdns/mdns.cpp

```diff
@@ -159,6 +147,164 @@ void Publisher::OnHostResolved(const std::string &aHostName, const Publisher::Di
     }
 }
 
+Publisher::SubTypeList Publisher::SortSubTypeList(SubTypeList aSubTypeList)
+{
+    std::sort(aSubTypeList.begin(), aSubTypeList.end());
+    return aSubTypeList;
+}
+
+Publisher::TxtList Publisher::SortTxtList(TxtList aTxtList)
+{
+    std::sort(aTxtList.begin(), aTxtList.end(),
+              [](const TxtEntry &aLhs, const TxtEntry &aRhs) { return aLhs.mName < aRhs.mName; });
+    return aTxtList;
+}
+
+std::string Publisher::MakeFullServiceName(const std::string &aName, const std::string &aType)
+{
+    return aName + "." + aType + ".local";
+}
+
+std::string Publisher::MakeFullHostName(const std::string &aName)
+{
+    return aName + ".local";
+}
+
+void Publisher::AddServiceRegistration(ServiceRegistrationPtr aServiceReg)
+{
+    mServiceRegistrations[MakeFullServiceName(aServiceReg->mName, aServiceReg->mType)] = aServiceReg;
+}
+
+void Publisher::RemoveServiceRegistration(const std::string &aName, const std::string &aType)
+{
+    otbrLogInfo("Removing service %s.%s", aName.c_str(), aType.c_str());
+    mServiceRegistrations.erase(MakeFullServiceName(aName, aType));
+}
+
+Publisher::ServiceRegistrationPtr Publisher::FindServiceRegistration(const std::string &aName, const std::string &aType)
+{
+    auto it = mServiceRegistrations.find(MakeFullServiceName(aName, aType));
+
+    return it != mServiceRegistrations.end() ? it->second : nullptr;
+}
+
+Publisher::ResultCallback Publisher::HandleDuplicateServiceRegistration(const std::string &aHostName,
+                                                                        const std::string &aName,
+                                                                        const std::string &aType,
+                                                                        const SubTypeList &aSubTypeList,
+                                                                        uint16_t           aPort,
+                                                                        const TxtList &    aTxtList,
+                                                                        ResultCallback &&  aCallback)
+{
+    ServiceRegistrationPtr serviceReg = FindServiceRegistration(aName, aType);
+
+    VerifyOrExit(serviceReg != nullptr);
+
+    if (serviceReg->IsOutdated(aHostName, aName, aType, aSubTypeList, aPort, aTxtList))
+    {
+        RemoveServiceRegistration(aName, aType);
+    }
+    else if (serviceReg->IsCompleted())
+    {
+        // Returns success if the same service has already been
+        // registered with exactly the same parameters.
+        std::move(aCallback)(OTBR_ERROR_NONE);
+    }
+    else
+    {
+        // If the same service is being registered with the same parameters,
+        // let's join the waiting queue for the result.
+        serviceReg->mCallback = std::bind(
```

Comment:
Instead of using std::bind, can we simply use a container like vector<ResultCallback> too allow multiple callbacks?

Reply:
It should work, but chaining callbacks with lambdas is just simpler and straightforward (it's also a little more code to handle a vector of callbacks :)